akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -319,8 +319,15 @@ private[stream] final class GraphInterpreterShell(
   private var self: ActorRef = _
   lazy val log = Logging(mat.system.eventStream, self)
 
-  lazy val interpreter = new GraphInterpreter(assembly, mat, log, inHandlers, outHandlers, logics,
-    (logic, event, handler) ⇒ self ! AsyncInput(this, logic, event, handler), settings.fuzzingMode)
+  private var canEnqueueToShortCircuit: (AsyncInput) ⇒ Boolean = _
+
+  lazy val interpreter: GraphInterpreter = new GraphInterpreter(assembly, mat, log, inHandlers, outHandlers, logics,
+    (logic, event, handler) ⇒ {
+      val asyncInput = AsyncInput(this, logic, event, handler)
+      val currentInterpreter = GraphInterpreter.currentInterpreterOrNull
+      if (currentInterpreter == null || (currentInterpreter.context ne self) || !canEnqueueToShortCircuit(asyncInput))
```

Comment:
Yes, but perhaps need to have canEnqueueToShortCircuit(from ActorGraphIntrpreter as it has queue of such async calls) function there in GraphIntrpreter